### PR TITLE
[Issue 1485] Interceptor on server app to swap REST URL with SSR REST URL

### LINKED ISSUE #1485: Unable to route network requests to different DSpace REST API URL during SSR
**Is your feature request related to a problem? Please describe.**

With Angular Universal, during server-side-rendering (SSR), REST requests are made from server-to-server. After rehydration the client-side-renderting (CSR) REST requests are made from the browser. Using the same URL for both is limiting for production infrastructure deployment causing server-to-server communication to route out of the internal network and back through the public URL which is often through load balancer or proxy. This decreases initial page load performance due to the increased network latency.

Additionally, running locally in a docker network requires either running dspace-angular in development mode or having a public IP for the host machine.

**Describe the solution you'd like**

The configuration should support providing both URL for DSpace API for when SSR makes requests and for when CSR makes requests. The implementation should alternate which URL to use based on `isPlatformBrowser()`.

**Describe alternatives or workarounds you've considered**

N/A

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by wwelling**
## References
* Fixes #1485

* Replaces https://github.com/DSpace/dspace-angular/pull/1541

## Description
Afford configuring REST URL for SSR to have direct server-to-server communication. Will allow for dspace-angular docker to use production build.

## Instructions for Reviewers
Configure SSR REST URL to other than REST URL.

Example for running in a docker network
```
      - DSPACE_SSR_REST_SSL=false
      - DSPACE_SSR_REST_HOST=dspace
      - DSPACE_SSR_REST_PORT=8080
      - DSPACE_SSR_REST_NAMESPACE=/server
```

REST API request during SSR will be to SSR REST and during CSR will be to REST.

List of changes in this PR:
* Add `ssr.rest` ServerConfig to AppConfig interface
* Default `ssr.rest` to `rest` ServerConfig
* Build `ssr.rest` base URL
* Add server app module REST interceptor to swap request URL rest.baseUrl with ssr.rest.baseUrl
* Add client app module REST interceptor to swap request URL ssr.rest.baseUrl with rest.baseUrl

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Comment by tdonohue** [comment]
Thanks @wwelling !  I'll give this a test on my end.  One quick note though, it looks like this is failing e2e tests right now.  It looks like maybe it's using the wrong URL to access the (Docker-based) backend that we spin up in GitHub CI?  (At least that's a first guess, I need to dig further though to understand this code... so I might be wrong as to what is happening)

**Review by tdonohue — COMMENTED** [review]
@wwelling :  I gave this a look today & a test.

Overall, the code makes sense to me, and I understand how it's supposed to work.  But, I've run into some immediate issues with how it works when I test it.  That said, the basics "work".. it's just buggy in some areas.

**First I did a basic test using a Docker-based backend.** Unfortunately though, it didn't work.
I cannot get the "ssr.rest.*" settings to work with a local Docker backend.  I tried setting up the default `rest.*` settings to go against http://localhost:8080/server/ (as usual), and then tried setting the `ssr.rest.*` to use http://dspace:8080/server (as my docker container is named "dspace").  But, I just get 500 errors implying the backend isn't found or the backend won't trust the frontend.

**Second, I tried the same Docker-based backend behind a proxy**, and it **mostly works**.  I used [localtunnel](https://theboroer.github.io/localtunnel-www/) to give my Docker-based backend a public URL (https://tim.loca.lt/server/).  Then I tried again and set the `rest.*` settings to use that public URL (https://tim.loca.lt/server/) while the `ssr.rest.*` settings use http://localhost:8080/server/   This time, the UI loads successfully, and I see the `SSR replace` logs and they look correct.

However, the issues start when the SSR switches to CSR.  There are no errors in the SSR logs, but once it switches to CSR, I get a *large number of errors* in my browser's Console.  These errors all appear the be related to _permissions checks_ and it's because the permission check request has an embedded URL as a `uri` param like this:

```
GET https://tim.loca.lt/server/api/authz/authorizations/search/object?uri=http://localhost:8080/server/api/core/sites/6d65c6a2-3fe7-44dd-bacb-79271257c35d&feature=isCollectionAdmin&embed=feature
```
The problem is the `uri` param is still using the `ssr.rest.*` URL even though it's trying to run from CSR.  This mix of URLs is not acceptable to the REST backend & it throws a 400 error saying the "Request is invalid" because "the supplied uri is not ours".

**Strangely though** I also realized that these `/api/authz/authorizations` requests appear to be done TWICE... once via SSR (where they work) and then *again* via CSR (where a 400 error is thrown).

In the SSR logs, I see that same permissions check & it succeeds
```
SSR request https://tim-dev.loca.lt/server/api/authz/authorizations/search/object?uri=http://localhost:8080/server/api/core/sites/6d65c6a2-3fe7-44dd-bacb-79271257c35d&feature=isCommunityAdmin&embed=feature
SSR replace https://tim-dev.loca.lt with http://localhost:8080
SSR request after http://localhost:8080/server/api/authz/authorizations/search/object?uri=http://localhost:8080/server/api/core/sites/6d65c6a2-3fe7-44dd-bacb-79271257c35d&feature=isCommunityAdmin&embed=feature
```

After that, when the site switches to CSR, the **same permission check runs a second time** but this time it fails (again because the URL in the `uri` param doesn't match the URL of the request).  I see this in the CSR logs in my browser console:
```
CSR request https://tim.loca.lt/server/api/authz/authorizations/search/object?uri=http://localhost:8080/server/api/core/sites/6d65c6a2-3fe7-44dd-bacb-79271257c35d&feature=isCollectionAdmin&embed=feature

# This is where the backend now returns a 400 exception in response
```

Overall, the code seems to **mostly work** & I feel it's really close.  There seems to be two ways to possibly solve this problem:
* If this permissions request only ran **once** (via SSR) then the problem would go away. (I'm stumped right now as to why this same permissions check runs twice.. it seems wasteful to run it via SSR and then again via CSR.)
* Or, we'd need to find a way to update the CSR interceptor to also perform replacements in URL parameters (like that `uri` param). 

A few more minor comments inline below.

File: src/config/default-app-config.ts [hunk 1]
```diff
@@ -57,6 +57,18 @@ export class DefaultAppConfig implements AppConfig {
     nameSpace: '/',
   };
 
+  // The SSR REST API server settings, defaulting to rest.
+  // NOTE: these must be include in rest.cors.allowed.origins in your backend's local.cfg.
```
[the remark this answers is not in the document]
Comment (wwelling) [reply src/config/default-app-config.ts:61]: An internal network URL could be `http://dspace-angular:4200` and the external web server URL could be `https://dspace.angular.edu`. Wont the UI server-side origin making requests to the API server-side from `http://dspace-angular:4200` while the browser origin making the requests from URL `https://dspace.angular.edu`?

File: src/app/core/csr-rest-url/csr-rest-url.interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+   *
+   * @param httpRequest
+   * @param next
+   */
+  intercept(httpRequest: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    console.log(`CSR request ${httpRequest.url}`);
+    if (this.from !== this.to && httpRequest.url.startsWith(this.from)) {
+      console.log(`CSR replace ${this.from} with ${this.to}`);
+      const url = httpRequest.url.replace(this.from, this.to);
+      console.log(`CSR request after ${url}`);
```
[the remark this answers is not in the document]
Comment (wwelling) [reply src/app/core/csr-rest-url/csr-rest-url.interceptor.ts:32]: Yes, once tested and vetted as an acceptable solution.

**Comment by wwelling** [comment]
@tdonohue I closed the original PR due to similar issues. It did not work as well as I expected in certain architectures. i.e., docker network.

File: src/app/core/ssr-rest-url/ssr-rest-url.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * SSR REST Interceptor intercepting Http Requests, switching base URL for server-to-server request
+ */
+export class SsrRestUrlInterceptor implements HttpInterceptor {
+
+  private from: string;
+  private to: string;
+
+  constructor(
+    @Inject(APP_CONFIG) protected appConfig: AppConfig,
+    @Inject(REQUEST) protected req: any
```
Comment (ybnd) [inline src/app/core/ssr-rest-url/ssr-rest-url.interceptor.ts:19]: This `req` is not used

File: src/config/app-config.interface.ts [hunk 4]
```diff
@@ -25,6 +25,9 @@ import { FilterVocabularyConfig } from './filter-vocabulary-config';
 interface AppConfig extends Config {
   ui: UIServerConfig;
   rest: ServerConfig;
+  ssr: {
```
Comment (ybnd) [inline src/config/app-config.interface.ts:28]: `BuildConfig` may be a more appropriate place for this

**Review by ybnd — CHANGES_REQUESTED** [review]
@wwelling @tdonohue

Have you noticed any "stuck loading animations" with this PR? 

I've had some issues with this on `/home` (recent submissions), `/community-list` and `/search`, but not consistently -- not sure if it's due to Angular or REST...

<br>

While looking into @tdonohue's issue above it occured to me that the current implementation introduces a potential security risk: it exposes the port the REST server is running on even if it's not included in `dspace.server.url`

You'd need to dig a bit into the transfer state JSON included in the SSR HTML, but the port is just there.

I think we can fix this by 

* excluding `ssr.rest.*` from `APP_CONFIG_STATE` [here](https://github.com/DSpace/dspace-angular/blob/485bb840ce7f8cb6e92537de4fba1b2f9e84de6d/src/modules/app/server-init.service.ts#L91-L91)

* making sure responses are cached by their "public URL"

  * This also ensures that resource links always match up with `dspace.server.url`, so the CSR interceptor won't be needed and the "supplied uri is not ours" errors won't come up anymore

  * I managed to bodge this together by replacing the URLs for all `_links` [here](https://github.com/DSpace/dspace-angular/blob/485bb840ce7f8cb6e92537de4fba1b2f9e84de6d/src/app/core/data/dspace-rest-response-parsing.service.ts#L89)

    ```typescript
    ...
    if (isNotEmpty(data)) {
      if (isNotEmpty(data._links)) {
        for (const property of Object.keys(data._links)) {
          data._links[property].href = data._links[property].href.replace(this.ssrUrl, this.url);
        }
      }
      ...
    ```

    After this the duplicate authorization requests don't fail anymore.

    **Just an example; this doesn't cover all response parser subclasses**

File: src/config/default-app-config.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tdonohue) [reply src/config/default-app-config.ts:61]: @wwelling : I might be missing your point, but `rest.cors.allowed.origins` is the list of *client* `Origin` headers which the REST API will allow to login, etc.

In this scenario, the SSR REST API settings are not a new *client*, they are just a different way of contacting the backend.  So, I don't think they result in the *client's* `Origin` having a different value.

However, you are correct that if the backend receives requests from two different clients (an external proxy URL and an internal SSR client), then both those clients may need to be added to `rest.cors.allowed.origins`.   However, I don't believe either of those clients will be using the new `ssr.rest.*` settings (as you said, the client will always either use the external URL or an internal server-side client ULR like http://dspace-angular:4000, and neither of those correspond to the `ssr.rest.*` settings).  

In summary, I don't believe the `ssr.rest.*` URL settings need to be added to `rest.cors.allowed.origins`, unless I'm missing your point.  But, you are correct that we may need to determine if the server-side client (like http://dspace-angular:4000) needs to be added to `rest.cors.allowed.origins` (I haven't tested that yet).

[the remark this answers is not in the document]
Comment (wwelling) [reply src/config/default-app-config.ts:61]: Ah, I see the miscommunication. Yes, the comment was intended to be the clients associated with two different URL to make request to the server-side API. The comment can go but should be documented somewhere because the angular SSR client can have a different origin than the URL provided for the CSR client.

**Comment by wwelling** [comment]
@ybnd thanks for the reminder. Those HATEOAS links were giving me problems initially as well. I believe there may need to be a better solution than this or possible a REST API addition.

File: src/config/app-config.interface.ts (hunk 4 above)
Comment (wwelling) [reply src/config/app-config.interface.ts:28]: I have no strong preference. @tdonohue where would you prefer?

**Comment by ybnd** [comment]
@wwelling @tdonohue the link issue could be tackled in REST

Right now, the REST responses you get can contain a mix & match of links starting from `dspace.server.url` and links starting from the URL you send the request to
* Most links start from the destination URL
* `/server/api` links _except for `self`_ start from `dspace.server.url`

Would it be a problem to make _all_ REST response links start from `dspace.server.url`?

**Comment by tdonohue** [comment]
@ybnd : I'm not sure if that solution for the REST API would actually work, as I think it could cause issues on the SSR site.  But, we could try it out to see what happens, I guess.

I'm just concerned that when SSR tries to "follow" `_links`, it would now encounter all URLS that start with `dspace.server.url`... essentially bypassing what this PR is trying to accomplish.

Essentially, the point of this PR is obviously to allow CSR and SSR to use different URLs.  But, if the REST API always responds only with `_links` that start with `dspace.server.url`, that essentially will tell the client to always use CSR-based URLs (as `dspace.server.url` corresponds to the URL that CSR always must use).

File: src/config/app-config.interface.ts (hunk 4 above)
Comment (tdonohue) [reply src/config/app-config.interface.ts:28]: I think we should find a way to get this fully working before we tweak it this PR much further.  Minor tweaks here don't matter too much until we can a solution we like. 

Arguably, these settings could also move entirely under the existing `universal` configs, because that's where other settings related to SSR currently reside.  So, we could restructure this like:

```
universal:
  preboot: true
  async: true
  time: false
  rest:
     ssl: true
     host: api7.dspace.org
     port: 443
     nameSpace: /server
```

That would clarify that these settings are directly related, rather than having different config sections for "universal" vs "ssr".

**Comment by ybnd** [comment]
@tdonohue the same interceptor would work for any `_links` Angular will try to follow during SSR, so that shouldn't be a problem AFAIK

I looked into the necessary REST changes yesterday: it's not super straightforward

**Comment by tdonohue** [comment]
As discussed in yesterday's Dev meeting, we are rescheduling this for 7.6 as there's not enough time to develop a better solution for 7.5.  I've added the "needs discussion" label because we believe we may need to brainstorm possible solutions in a future dev meeting.
